### PR TITLE
Deepspeed pipeline parallel and fairscale sharded optimizer test samples with ORTModule

### PR DESCRIPTION
**Pull request by harshithapv**
Adding samples for Deepspeed pipeline parallel and fairscale sharded optimizer with ORTModule.
This PR is re-submitted from https://github.com/microsoft/onnxruntime/pull/6562 after rebasing with latest and address comments from old PR.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                        help='distributed backend')
+    parser.add_argument('--seed', type=int, default=0, help='PRNG seed')
+    parser.add_argument('--fp16',type=bool,default=False,help='fp16 run')
+    parser.add_argument('--run_without_ort',type=bool,default=False,help='onlydeepspeed run')
+    
+    parser = deepspeed.add_config_arguments(parser)
+    args = parser.parse_args()
+    return args
+
+print('Running deepspeed pipeline parallel module with ORTModule')
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:55]:
```suggestion
if args.run_without_ort:
   print('Running deepspeed pipeline parallel module without ORTModule')
else:
   print('Running deepspeed pipeline parallel module with ORTModule')
```
This test can be execute with and without ortmodule, right?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+n = 10
+d_in = 4
+d_hidden = 8
+d_out = 3
+import sys
+print(sys.version)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:61]:
```suggestion
```
remove debug

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        nn.ReLU(),                     # Stage 2
+        nn.Linear(d_hidden, d_out)     # Stage 2
+    )
+
+
+else:
+    print("*************************************** testing with ORTModule ************")
+    model = nn.Sequential(
+        ORTModule(nn.Linear(d_in, d_hidden).to(device)),     # Stage 1
+        nn.ReLU().to(device),#ORTModule(nn.ReLU().to(device)),                     # Stage 1
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:86]: Add a TODO comment saying ORTModule can wrap Relu once stateless model is supported. @baijumeswani is working on it

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+print("Initialize deepspeed")
+model_engine, optimizer, _, _ = deepspeed.initialize(args=args,
+                                                     model=model,
+                                                     model_parameters=params,
+                                                     training_data=ds #(x,y)#
+                                                    )
+
+for step in range(args.steps):
+    loss = model_engine.train_batch()
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:119]: Does `model_engine.train_batch()` print loss or any other metric

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                        help='how many batches to wait before logging training status (default: 300)')
+    parser.add_argument("--cpu", action="store_true", default=False)
+    parser.add_argument("--use_ortmodule", action="store_true", default=False, help = "use ortmodule")
+    parser.add_argument('--test-batch-size', type=int, default=64, metavar='N',
+                        help='input batch size for testing (default: 64)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--seed', type=int, default=42, metavar='S',
+                        help='random seed (default: 42)')
+    parser.add_argument('--data_dir', type=str, default='./mnist',
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py:253]:
```suggestion
    parser.add_argument('--data-dir', type=str, default='./mnist',
```
The argument is passed as `--data-dir` from `orttraining_ortmodule_distributed_tests.py:run_ortmodule_fairscale_sharded_optimizer_tests`

**Review by thiagocrepaldi — COMMENTED** [review]
How much time does this test add to the total pipeline run?

**Comment by harshithapv** [comment]
> How much time does this test add to the total pipeline run?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py (hunk 5 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_fairscale_sharded_optimizer.py]: addressed this.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 4 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:122]: it prints the final loss. But the latest commit  prints loss every 10 steps.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 3 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py]: done

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 2 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py]: done

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 1 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py]: done

Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py]: yes it can run without ORTModule and matches the result without ORT. Also, Addressed your suggestions. Thank you.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py (hunk 2 above)
Comment (harshithapv) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py]: done

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        nn.Linear(d_hidden, d_hidden), # Stage 1
+        nn.ReLU(),                     # Stage 1
+        nn.Linear(d_hidden, d_hidden), # Stage 2
+        nn.ReLU(),                     # Stage 2
+        nn.Linear(d_hidden, d_out)     # Stage 2
+    )
+
+
+else:
+    print("*************************************** testing with ORTModule ************")
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_deepspeed_pipeline_parallel.py:86]: this is the same log as line 66 :)

**Review by thiagocrepaldi — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_ortmodule_distributed_tests.py [hunk 7]
```diff
@@ -33,6 +33,22 @@ def run_ortmodule_deepspeed_zero_stage_1_tests(cwd, log, data_dir):
...
+    log.debug('Running: ORTModule deepspeed pipeline parallel tests')
+
+    command = ['deepspeed', 'orttraining_test_ortmodule_deepspeed_pipeline_parallel.py',
+        '--deepspeed_config', 'orttraining_test_ortmodule_deepspeed_pipeline_parallel_config.json']
+
+    run_subprocess(command, cwd=cwd, log=log).check_returncode()
+
+def run_ortmodule_fairscale_sharded_optimizer_tests(cwd, log, data_dir):
+    log.debug('Running: ORTModule fairscale sharded optimizer tests')
+    command = ['python3', 'orttraining_test_ortmodule_fairscale_sharded_optimizer.py',
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_ortmodule_distributed_tests.py:46]: We could use sys.executable instead. So that it runs with the same python version as the current file is being run in.

File: tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage2/requirements.txt [hunk 8]
```diff
@@ -1 +1,2 @@
 deepspeed
+fairscale
```
Comment (baijumeswani) [inline tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage2/requirements.txt:2]: can this be a part of stage1?

**Comment by g-karthik** [comment]
@harshithapv somewhat unrelated question, but is there a doc that describes how to use ORT with DeepSpeed?

I have a script that uses DeepSpeed (stages 1 and 2) to train a Hugging Face model and I would like to modify it to use ORT so that training is faster.

File: tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage2/requirements.txt (hunk 8 above)
Comment (harshithapv) [reply tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage2/requirements.txt:2]: fairscale has a dependency on torch. So will put this under stage 2

File: orttraining/orttraining/test/python/orttraining_ortmodule_distributed_tests.py (hunk 7 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_ortmodule_distributed_tests.py:46]: Can address this in the next PR.
